### DISCUSSION TITLE
A way to disconnect LAN mode in order to use Bambu Studio

### DISCUSSION THREAD

**Discussion by EnigIVIa** [greghesp/ha-bambulab#301] [Q&A]
Is there a way to disconnect LAN mode in order to use Bambu Studio. Since the P1S only allows one LAN connection as soon as I setup the homeassistant intergration (which is amazing) I cant start prints anymore since bambu studio cant connect to the printer. Is there a disconnect button or the possibility to connect to the printer only every 5 minutes or so to give bambu studio a chance to connect?

**Comment by AdrianGarside** [comment]
I added back the ability to connect via bambu cloud. I currently have no plans to further workaround the bug Bambu introduced in the P1* firmware. See #174 for the link to their bug tracking the regression.

**Comment by AdrianGarside** [comment]
Can you give this a go to see if it's workable for you? https://github.com/greghesp/ha-bambulab/releases/tag/v1.9.3-dev3-manualrefresh

**Reply by EnigIVIa** [reply]
First of all thanks for working on that feature, I very much appreciate that. I tested it briefly and will give it some more time the upcoming weekend. Pressing the "Manual refresh mode" button works in the sense that after pressing it I'm again able to connect with Bambu Studio. What's a little bit confusing is that this button shows up as some kind of push-button rather than a toggle-button. So I never know if I'm in manual mode or not.

**Reply by AdrianGarside** [reply]
The icon changes depending on the state. I’ll see if I can switch it to a checkbox style instead as that would be clearer still.

**Reply by AdrianGarside** [reply]
https://github.com/greghesp/ha-bambulab/releases/tag/v1.9.3-dev3-manualrefresh2

**Reply by EnigIVIa** [reply]
Gave it a quick spin. Button looks good now but switches automatically back to off if I press it.

**Reply by AdrianGarside** [reply]
Can you take a video? It’s just a switch now and there’s no code to automate setting it except for the initial off state.

**Reply by EnigIVIa** [reply]
Sure:

https://github.com/greghesp/ha-bambulab/assets/2556777/01cd0066-3676-4d76-baa0-f156a7bf18ed

**Reply by AdrianGarside** [reply]
That's really bizarre. Are there any errors in the log when you see that?

**Reply by AdrianGarside** [reply]
Also per this, there's a chance the firmware bug will be fixed in the upcoming 1.5 firmware update: https://github.com/bambulab/BambuStudio/issues/2404#issuecomment-1857378019

**Reply by AdrianGarside** [reply]
One thing I notice in your video is that the printer is reporting as offline. It shouldn't be with this approach. So I suspect that in this case this instance of home assistant is running into the bug - it's unable to get data from the printer. Although I still have no theory for how the switch can turn itself off since there's literally no code to do so.

**Reply by EnigIVIa** [reply]
Nice to hear that bambu lab claim they will fix it. That would be the beste solution. I indeed got an error in the logs:

This error originated from a custom integration.

Logger: custom_components.bambu_lab.pybambu
Source: custom_components/bambu_lab/pybambu/bambu_client.py:280
Integration: Bambu Lab (documentation, issues)
First occurred: 9:58:06 AM (1 occurrences)
Last logged: 9:58:06 AM

On Disconnect: Disconnected from Broker: 0

**Reply by AdrianGarside** [reply]
That's expected - when you turn on manual mode it disconnects and that's currently logged as a warning level log. I still have no theory as to how you can get the behavior you're seeing.

**Reply by EnigIVIa** [reply]
I tested it again an the bug also happens when the printer is in idle state. It seems like it always stays in manual mode. I also had a look at the code and get why this behavior sounds impossible. My guess is that this is some weird timing issue with the frontend. Since the button shows the state but is also a control element at the same time there might be something going on with the state of the button being set to the current value of the backend prior to doing the switching. The time until it switches back is about 2 seconds, maybe that matches some refresh interval of the homeassistant frontend.

**Reply by AdrianGarside** [reply]
I accidentally left the button enabled for X1 (instead of just P1/A1 as intended). And my X1 is currently in a broken state in my production instance of HA where it's not online according to Bambu. (A problem I having from time to time with my X1 that Bambu support wants me to 'solve' by factory reseting it and readding it to my account which I'm highly skeptical will help). And I repro the same behavior you see. The attempt to turn the manual mode on doesn't work and gets reset a moment later. When you see this - is the data that HA currently has for the printer correct / up to date?

**Reply by AdrianGarside** [reply]
Even with the local repro I couldn't learn anything - that instance of the local mqtt connection wasn't being given anything other than the version info on reconnection. So it seems like a firmware issue - likely related to the fact the printer is offline according to bambu cloud. Restarting home assistant fixed both the inability to receive data and the behavior of the manual refresh button.

I think your repro is related - I suspect it happens when the HA instance is already blocked from receiving data from the local mqtt by the firmware bug. In my testing it wasn't as simple as 'last connection wins' as disconnecting and reconnecting a 'no-data' instance did not make it the last one.

**Reply by EnigIVIa** [reply]
I'm sorry to hear that this feature is causing you trouble. Your are correct, a restart of HA fixed it and the button is behaving normal now. I created an automation that puts the integration in manual refresh mode after a print is finished. That should fix the issue for me. Thanks again for taking that on. I'm very excited about the integration, now I can even get push messages after a print is finished without the online mode :-)

**Reply by AdrianGarside** [reply]
I think this might fix the problem you are seeing: https://github.com/greghesp/ha-bambulab/releases/tag/v1.9.3-dev3-manualrefresh3
And I suspect you'll need this for reliable behavior even with the automation approach you've setup.

**Reply by AdrianGarside** [reply]
I was thinking you could also use automation to create a polling mode with the new support. I originally thought about building that into the new support but I think you can achieve it without and it was much simpler to just make this workaround all user (/automation) initiated.

**Reply by EnigIVIa** [reply]
Ah you mean setting the integration to manual refresh and hitting the refresh button with an automation every x seconds/minutes. But what will happen if it triggers while bambu studio is active? Can't that lead to HA being locked out again?

**Reply by AdrianGarside** [reply]
Bambu Studio will briefly not receive updates for the ~3s the HA integration is connected. So there’s a small chance it might miss something important like turning the chamber light on/off but I think it’ll mostly work acceptably.

**Comment by AdrianGarside** [comment]
FYI that the beta  01.04.50.19 firmware appears to have fixed the mqtt connection issue.
